### PR TITLE
Implementing reverse routing syntax

### PR DESCRIPTION
**Pull request by burzum**
See these two links

* https://github.com/cakephp/cakephp/issues/10643#issuecomment-301652587
* https://github.com/cakephp/cakephp/issues/10689#issuecomment-304440165

### CODE REVIEW

File: src/Routing/RouteBuilder.php [hunk 1]
```diff
@@ -703,8 +703,12 @@ public function loadPlugin($name, $file = 'routes.php')
      * @throws \InvalidArgumentException
      * @throws \BadMethodCallException
      */
-    public function connect($route, array $defaults = [], array $options = [])
+    public function connect($route, array $defaults = null, array $options = [])
```
Comment (ADmad) [inline src/Routing/RouteBuilder.php]: Why change the default value?

File: src/Routing/Router.php [hunk 2]
```diff
@@ -202,14 +202,59 @@ public static function getNamedExpressions()
...
+                        'controller' => $matches[3],
+                        'view' => $matches[4]
+                    ];
+                case 4:
+                    return [
+                        'plugin' => $matches[1],
+                        'prefix' => $matches[2],
+                        'controller' => $matches[3],
+                        'view' => $matches[4]
+                    ];
```
Comment (markstory) [inline src/Routing/Router.php]: There should probably be an exception thrown for the default case.

File: tests/TestCase/Routing/RouterTest.php [hunk 3]
```diff
@@ -3422,6 +3422,53 @@ public function testCreateRouteBuilder()
...
+            'controller' => 'Articles',
+            'view' => 'view',
+            'action' => 'index',
+            'plugin' => null,
+            '_matchedRoute' => '/admin/articles/view'
+
+        ];
+        $this->assertEquals($result, $expected);
+
+        Router::connect('/admin/blog/articles/view', 'Blog.Admin\Articles::view');
```
Comment (markstory) [inline tests/TestCase/Routing/RouterTest.php]: What about nested prefixes? Those have caused us troubles in other places.

Comment (markstory) [reply tests/TestCase/Routing/RouterTest.php]: Or all tokens separated by \ except the last one are the prefix, and the last one is the controller name.

Comment (burzum) [reply tests/TestCase/Routing/RouterTest.php]: @markstory like that?

image

Comment (markstory) [reply tests/TestCase/Routing/RouterTest.php]: Is `Foo\Bar\Admin` the prefix? Or the plugin name there?

Comment (burzum) [reply tests/TestCase/Routing/RouterTest.php]: Prefix? Just specific exactly how you would like to have it implemented? :) Can you just commit a test case for the syntax you want and I'll take care of the rest.

**Comment by markstory** [comment]
@burzum I've done the changes I think needed to be done and pushed them up.

**Comment by burzum** [comment]
@markstory looks good.

**Comment by markstory** [comment]
I will merge this once I get the documentation updates done.

**Pull request merged**